feat: support filtering for warnings only on the review page

fhennig opened the pull request:
<!-- Mention the issue that this PR resolves. Delete if there is no corresponding issue -->
part of #3031

fixes #2566
fixes #1447 
fixes #3198 -- [ac6aa79](https://github.com/loculus-project/loculus/pull/3125/commits/ac6aa79fd22d27abdc3e2b67feb5f31e9fc109c8)

<!-- Add "preview" label in almost all cases to have testable deployment. Then lookup the deployment URL in argocd and paste it here (if you don't know how to look up the URL, ask here: https://loculus.slack.com/archives/C06JCAZLG14), it's something like `{REPLACE}.loculus.org` -->
preview URL: https://feat-3031-improve-review.loculus.org/

### Summary
<!-- Add a few sentences describing the main changes introduced in this PR -->
This PR merges the HAS_ERRORS and AWAITING_APPROVAL statuses in the SequenceEntryView. Instead errors are handled by looking at the errors column. The `get-sequences` endpoint now supports filtering by processing result (`no_issues`, `warnings`, `errors`) and also returns the counts for these states.

The UI has been adapted to support filtering.

Documentation has been updated.

The PR also adds confirmation dialogs to individual sequence publishing or deletion (#2566)

### Screenshot
<!-- When applicable, add a screenshot showing the main effect this PR has, even if "trivial": e.g. changed layout, changed button text, different logging in backend. This helps others quickly grasping what you did even if they are not familiar with the code base. -->
<img width="1379" alt="image" src="https://github.com/user-attachments/assets/b2720c34-0926-4629-b312-2188e4d585c2">


### PR Checklist
<!-- Check completed items of strikethrough irrelevant items (using ~~text~~) -->
- [ ] All necessary documentation has been adapted.
- [ ] The implemented feature is covered by an appropriate test.

fhennig commented:
Hmm, if I run `ktlintMainSourceSetCheck` locally it's all green 🤔

theosanderson commented on backend/src/main/kotlin/org/loculus/backend/api/SubmissionTypes.kt:
```diff
@@ -43,6 +43,16 @@ enum class WarningsFilter {
 }
 
+@Schema(
+    description = "If set to 'INCLUDE_ERRORS', sequence entries with warnings are included in the response." +
```
```suggestion
    description = "If set to 'INCLUDE_ERRORS', sequence entries with errors are included in the response." +
```
and below

theosanderson commented on backend/src/main/kotlin/org/loculus/backend/controller/SubmissionController.kt:
```diff
@@ -347,6 +348,7 @@ open class SubmissionController(
         ) @RequestParam(required = false) statusesFilter: List<Status>?,
         @HiddenParam authenticatedUser: AuthenticatedUser,
         @RequestParam(required = false, defaultValue = "INCLUDE_WARNINGS") warningsFilter: WarningsFilter,
+        @RequestParam(required = false, defaultValue = "INCLUDE_ERRORS") errorsFilter: ErrorsFilter,
```
So I think given the UI we will want to filter in the same way as we do the counting - including or not "PERFECT" or "WARNINGS" or "ERRORS"

fhennig replied:
The API has now changed to include a processingResultFilter where you can include (or not) these states

corneliusroemer commented:
> Hmm, if I run ktlintMainSourceSetCheck locally it's all green 🤔

I get errors when running `./gradlew ktlintCheck` (unfold details)

<details>

```
loculus/backend on  feat/3031-improve-review-page [$?] is 📦 v0.0.1 via ☕ v21.0.4 via 🅶 v8.10.2 in ☸ aws (update-loculus-4f8ecf) on ☁️  loculus.eks (eu-central-1)
❯ ./gradlew ktlintCheck

> Task :ktlintMainSourceSetCheck FAILED
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/controller/SubmissionController.kt:24:1 Unused import
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/controller/SubmissionController.kt:35:1 Unused import
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:21:1 Unused import
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:45:1 Unused import
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:62:1 Unused import
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:693:25 Missing spacing after "{"
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:693:43 Newline expected after arrow
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:693:43 Missing newline after "->"
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:693:45 Missing newline after '->'
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:693:49 Missing spacing after "when"
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:697:17 Missing newline before "}"
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:697:18 Missing spacing after "}"
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:697:18 Missing spacing before "}"
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:697:18 Missing newline before '}'
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:697:18 Newline expected before closing brace
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:698:28 Missing spacing around "{"
/Users/corneliusromer/code/loculus/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:698:63 Missing spacing before "}"

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':ktlintMainSourceSetCheck'.
> A failure occurred while executing org.jlleitschuh.gradle.ktlint.worker.ConsoleReportWorkAction
   > KtLint found code style violations. Please see the following reports:
     - /Users/corneliusromer/code/loculus/backend/build/reports/ktlint/ktlintMainSourceSetCheck/ktlintMainSourceSetCheck.txt

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 6s
4 actionable tasks: 4 executed
```

</details>

theosanderson commented on website/src/hooks/useSubmissionOperations.ts:
```diff
@@ -25,8 +27,11 @@ export function useSubmissionOperations(
+    const allProcessingResults = [noIssuesProcessingResult, warningsProcessingResult, errorsProcessingResult];
     const [includedStatuses, setIncludedStatuses] = useState<string[]>(allRelevantStatuses);
+    const [includedProcessingResults, setIncludedProcessingResults] = useState<string[]>(allProcessingResults);
+    // TODO here I should add the include/exclude warnings thing
```
Is this still a todo or just needs deleting?

fhennig replied:
just needed deleting, thanks!

https://github.com/loculus-project/loculus/pull/3125/commits/01a6a5db34ea0dca22eb32aabfea5793d4d385b3

theosanderson reviewed (commented):
This looks really nice to me! Thanks so much. I'm not a good person to review Kotlin - if @chaoran-chen or @fengelniederhammer had time to look that might be helpful.

As discussed on Slack we can iterate on the exact UI but this seems really great for the moment

corneliusroemer commented on backend/src/main/kotlin/org/loculus/backend/api/SubmissionTypes.kt:
```diff
@@ -213,12 +203,18 @@ enum class PreprocessingAnnotationSourceType {
     override val accession: Accession,
     override val version: Version,
     val status: Status,
+    val isError: Boolean,
```
Actually IntelliJ has a refactor function that means it's 2 clicks instead of manual find/replace :)

fhennig replied:
thanks for pointing that out! Yes, not hard at all and I like being precise with these property names if possible!

fhennig replied:
Done: https://github.com/loculus-project/loculus/pull/3125/commits/6f7c757d7ce93cfa04f72906424f6d41eb474c3f

corneliusroemer commented on backend/src/main/kotlin/org/loculus/backend/controller/SubmissionControllerDescriptions.kt:
```diff
@@ -53,10 +53,10 @@ const val SUBMIT_PROCESSED_DATA_DESCRIPTION = """
- The user can then edit the data and submit a corrected version.
+will roll back all previously inserted data. It is the responsibility of the processing pipeline to ensure that the 
+content of the data is correct. If the pipeline is unable to provide valid data, it should submit the data with errors.
+In this case, no validation will be performed and the status of the accession version will be set to 'HAS_ERRORS'.
```
There are also some internal docs that use those two terms, I added that to the acceptance criteria earlier today https://github.com/loculus-project/loculus/issues/3031#issuecomment-2449304609

fhennig replied:
This particular code area I just saw that the whitespace was messy, so I fixed it. `HAS_ERRORS` still exists as a status on the Pipeline output (would have to look up the exact table) so I didn't replace it here. The change was just done in the view.

I also did a search for both terms and I think I got everything!

theosanderson replied:
Thanks - I hadn't understood that HAS_ERRORS existed in two different places. I suspect we may no longer need it. (I.e. it may is presumably redundant with the actual errors themselves, which are I presume what we are using) But (A) I may be wrong (B) it doesn't matter much.

fhennig replied:
I briefly touched on this with Anya and she said it's probably best to just leave the pipeline code as-is for now

anna-parker replied:
Sorry think I missed some context here - where does HAS_ERRORS exist in the pipeline output?

theosanderson replied:
Some investigation: it's not directly in the pipeline output, it's computed by the backend here: https://github.com/loculus-project/loculus/blob/ee8a842ff9a403cbe062cabd632cd3792974d5f0/backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt#L311

fhennig replied:
Huh, okay! I was always just working on this view, and just saw that the HAS_ERRORS is still there in a different table, so I thought maybe it's used there also for a bunch of things. But if we can get rid of it that would be good I think!

fhennig replied:
Who can assist me with this @chaoran-chen ? I'm not sure but I agree that it kind of looks like we don't actually needs this status at all anymore?

fhennig replied:
In the latest version of this PR, this was removed.

theosanderson commented on backend/src/main/resources/db/migration/V1.5__update_sequence_entries_view_status_field.sql:
(after you rebase and add the `update_db_schema` label this will generate a diff)

theosanderson replied:
Looks good to me. I guess it's not ideal we call revocations `processed` when they don't actually get processed but I can't think of something better

chaoran-chen commented on backend/src/main/kotlin/org/loculus/backend/api/SubmissionTypes.kt:
Would you maybe like to update https://github.com/loculus-project/loculus/blob/main/backend/docs/runtime_view.md? The file unfortunately is already quite outdated but it seems useful to me to have some docs on how data flows through the backend and now might be a good opportunity to update it again.

fhennig replied:
I don't _want_ to do it, but I can see how now would be a good time. I have it on my list for tomorrow

fhennig replied:
https://github.com/loculus-project/loculus/pull/3125/commits/46349888040602b69697a0285e79fe35f796e90d

chaoran-chen commented on backend/src/main/kotlin/org/loculus/backend/api/SubmissionTypes.kt:
Could you please update the Approve submisisons API docs [here](https://github.com/loculus-project/loculus/blob/a7952647e8a1c193ea27de53b470c3cff6e3c5ca/docs/src/content/docs/for-users/approve-submissions.md?plain=1#L39) (which I just moved from Pathoplexus a few days ago - sorry!) and later, when it is about to be used for Pathoplexus, also [here](https://github.com/pathoplexus/pathoplexus/blob/30fa22b8b4d87efc21fe16e52e00e98d0e4bc7d2/monorepo/website/src/pages/docs/how-to/approve-submissions.mdx?plain=1#L48).

fhennig replied:
https://github.com/loculus-project/loculus/pull/3125/commits/25bae87475e7a0172865e283b5423a3bc94f66a6

fhennig replied:
Is that good?

fhennig replied:
If that change is fine I'd make a similar PR for the ppx repo

chaoran-chen commented on backend/src/main/kotlin/org/loculus/backend/service/submission/AccessionPreconditionValidator.kt:
```diff
@@ -152,6 +155,21 @@ class AccessionPreconditionValidator(
 
+        fun andThatSequenceEntriesHaveNoErrors(): CommonPreconditions {
+            val sequenceEntriesWithErrors = sequenceEntries
+                .filter { row -> row[SequenceEntriesView.errorsColumn].orEmpty().isNotEmpty() }
```
```suggestion
                .filter { it[SequenceEntriesView.errorsColumn].orEmpty().isNotEmpty() }
```

chaoran-chen commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -481,7 +486,7 @@ open class SubmissionDatabaseService(
         }
 
         val scopeCondition = if (scope == ApproveDataScope.WITHOUT_WARNINGS) {
-            not(SequenceEntriesView.entriesWithWarnings)
+            not(SequenceEntriesView.hasWarnings)
```
Would it make sense to rename the variable to `warnignCondition`?

fhennig replied:
I have now made this particular condition private and refactored that whole bit to maybe make it clearer what's going on.

https://github.com/loculus-project/loculus/pull/3125/commits/59f73282057b80b8a297cb4a314f94b15129c208

chaoran-chen commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -621,12 +626,18 @@ open class SubmissionDatabaseService(
-        groupIdsFilter: List<Int>?,
-        statusesFilter: List<Status>?,
-        warningsFilter: WarningsFilter? = null,
+        organism: Organism? = null,
```
The nullability was introduced by Luke because this function is also called in the SeqSetCitationsController where the organism doesn't matter.

theosanderson replied:
Thanks

chaoran-chen commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -677,37 +690,97 @@ open class SubmissionDatabaseService(
+                // Filter for has errors or not, but only if status is processed
+                not(SequenceEntriesView.statusIs(Status.PROCESSED)) or
+                    (
+                        SequenceEntriesView.statusIs(Status.PROCESSED) and
```
```suggestion
```

This condition is redundant, right?

fhennig replied:
I think it is not entirely redundant, I have now a function like this:

```kotlin
    fun processingResultIs(processingResult: ProcessingResult) = when (processingResult) {
        ProcessingResult.ERRORS -> SequenceEntriesView.hasErrors
        ProcessingResult.NO_ISSUES -> not(
            SequenceEntriesView.hasErrors or SequenceEntriesView.hasWarnings,
        ) and SequenceEntriesView.statusIs(Status.PROCESSED)
        ProcessingResult.WARNINGS -> not(SequenceEntriesView.hasErrors) and
            SequenceEntriesView.hasWarnings
    }
```

I think for the NO_ISSUES path it is still relevant, otherwise we would lump unprocessed and processed but no issues together.

https://github.com/loculus-project/loculus/pull/3125/commits/59f73282057b80b8a297cb4a314f94b15129c208

chaoran-chen commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -677,37 +690,97 @@ open class SubmissionDatabaseService(
+                not(SequenceEntriesView.statusIs(Status.PROCESSED)) or
+                    (
+                        SequenceEntriesView.statusIs(Status.PROCESSED) and
+                            it.map { processingResult ->
```
This is very concise but it also took me quite a while to understand the whole block (starting with `processingResultFilter?.let {`). Could we somehow simplify it a bit?

fhennig replied:
I have untangled that whole code block and it should now be much clearer: https://github.com/loculus-project/loculus/pull/3125/commits/59f73282057b80b8a297cb4a314f94b15129c208

chaoran-chen commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -677,37 +690,97 @@ open class SubmissionDatabaseService(
+        authenticatedUser: AuthenticatedUser,
+        organism: Organism?,
+    ): Map<ProcessingResult, Int> {
+        val countBaseQuery = SequenceEntriesView
+            .select(
+                SequenceEntriesView.accessionColumn,
+                SequenceEntriesView.versionColumn,
+                SequenceEntriesView.errorsColumn,
+                SequenceEntriesView.warningsColumn,
+            )
+            .where { getGroupCondition(groupIdsFilter, authenticatedUser) }
+            .andWhere { SequenceEntriesView.statusIs(Status.PROCESSED) }
+
+        if (organism != null) {
+            countBaseQuery.andWhere { SequenceEntriesView.organismIs(organism) }
+        }
+
+        val errorCount = countBaseQuery.copy().andWhere { SequenceEntriesView.hasErrors }.count()
+        val warningCount = countBaseQuery.copy().andWhere {
+            SequenceEntriesView.hasWarnings and not(SequenceEntriesView.hasErrors)
+        }.count()
+        val noIssuesCount = countBaseQuery.copy().andWhere {
+            not(SequenceEntriesView.hasWarnings or SequenceEntriesView.hasErrors)
+        }.count()
+        val processingResultCounts = mapOf(
+            ProcessingResult.ERRORS to errorCount.toInt(),
+            ProcessingResult.WARNINGS to warningCount.toInt(),
+            ProcessingResult.NO_ISSUES to noIssuesCount.toInt(),
         )
```
A group by query would be more efficient and require one instead of three queries. Maybe like this?

```kotlin
val processingResultType = case()
    .When(SequenceEntriesView.hasErrors, stringLiteral("HAS_ERRORS"))
    .When(SequenceEntriesView.hasWarnings, stringLiteral("HAS_WARNINGS"))
    .Else(stringLiteral("NO_ISSUES"))

val countQuery = SequenceEntriesView
    .select(processingResultType, Count(stringLiteral("*")))
    .where { getGroupCondition(groupIdsFilter, authenticatedUser) }
    .andWhere { SequenceEntriesView.statusIs(Status.PROCESSED) }

if (organism != null) {
    countQuery.andWhere { SequenceEntriesView.organismIs(organism) }
}

countQuery.groupBy(processingResultType)
…
```

fhennig replied:
https://github.com/loculus-project/loculus/pull/3125/commits/942774ab09e7cf0810b82ad2cd236c47c4b2db58

Done!

fhennig replied:
(Good suggestion by the way, thanks!)

chaoran-chen reviewed (commented):
Thanks, Felix! I reviewed the Kotlin main code (not the tests and website code) and it looks good! Just a few small suggestions for improvements.

fhennig commented on website/src/components/ReviewPage/ReviewCard.tsx:
```diff
@@ -109,79 +111,59 @@ const ButtonBar: FC<ButtonBarProps> = ({
-                        ? 'Cannot edit. Wait for preprocessing!'
-                        : 'Edit this sequence entry'
-                }
+                content={notProcessed ? 'Cannot edit. Wait for preprocessing!' : 'Edit this sequence entry'}
```
I did "..." instead of ".." I hope that's fine? 

https://github.com/loculus-project/loculus/pull/3125/commits/2a22284ef44114d8b3c4ef6930a9844fc1beee4c

theosanderson commented:
Huh interesting the migration SQL thing may not be functioning as we expected, apologies. Away for a bit but will look into it.

fhennig commented:
Hmm, I see e2e tests fail on main I think? I'm not sure the e2e test failures are related to my changes in this PR

fhennig commented:
This is still TODO: https://github.com/loculus-project/loculus/pull/3125#discussion_r1826033855

fhennig commented:
I have now moved the processing_result into a column in the view, so it's not calculated in the backend code anymore. I have also updated the preprocessing status field to not include FINISHED and HAS_ERRORS anymore.

fhennig commented:
This is still TODO: https://github.com/loculus-project/loculus/pull/3125#discussion_r1826033855

fengelniederhammer commented on website/src/components/ReviewPage/ReviewCard.tsx:
```diff
@@ -48,6 +49,8 @@ export const ReviewCard: FC<ReviewCardProps> = ({
     accessToken,
 }) => {
     const { isLoading, data } = useGetMetadataAndAnnotations(organism, clientConfig, accessToken, sequenceEntryStatus);
+    // TODO the hook above makes an extra call for each row.
```
Is there an issue that we can reference in the TODO? (Otherwise TODOs will likely never be done)

fhennig replied:
yes there is a ticket for it, I'll just remove the TODO here: https://github.com/loculus-project/loculus/pull/3125/commits/6477e6dac5f33197c2b472d699350ecd99747ee9

fengelniederhammer reviewed (commented):
The backend part looks reasonable (I didn't check all details). And I just had a quick look over the website code.

fhennig commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -677,39 +677,79 @@ open class SubmissionDatabaseService(
+                    status = Status.fromString(row[SequenceEntriesView.statusColumn]),
+                    processingResult = if (row[SequenceEntriesView.processingResultColum] !=
+                        null
+                    ) {
```
agreed; fixed: https://github.com/loculus-project/loculus/pull/3125/commits/b7bc02c0a6645ec6b6991c0e5436f56a013ff473

fhennig commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -210,19 +216,7 @@ open class SubmissionDatabaseService(
             statusByAccessionVersion[submittedProcessedData.displayAccessionVersion()] = newStatus
```
simplified: https://github.com/loculus-project/loculus/pull/3125/commits/487651e7c526f15fdae37fcadaf35217e9ff50e7

fhennig commented:
Alright! Thanks for @corneliusroemer and @anna-parker pointing out that there wasn't any migration for the `processing_status` column, I have fixed that now! 

I have tested the migration locally by first deploying the DB from main, and then manually migrating it and deploying the new frontend and backend, and it was working for me.

corneliusroemer commented:
Thanks @fhennig, did you ensure that there was some data in the db? If there's nothing processed it could be a false positive. Easiest might actually to test this on staging with prod data. We should probably do this before merging in case fixes are necessary. And given the wide ranging changes there's a non-zero chance we'll hit some edge case that's hard to get with test data.

corneliusroemer commented:
Need to fix the merge conflicts before we can test this on staging unfortunately - when merge conflicts exist the CI doesn't run.

The CI runs on the outcome of the eventual merge, so with conflict it can't run.

corneliusroemer commented:
Great work, just reporting my staging.ppx.bio test results (where rollout worked!):

I think I found a bug, if you uncheck errors, one still gets shown a lot of pages to page through, even though there are _0_ results.

<img width="1196" alt="Google Chrome Beta 2024-11-11 19 15 10" src="https://github.com/user-attachments/assets/1a4ea12e-6d92-4053-8997-094e3a142f40">

theosanderson commented:
> if you uncheck errors, one still gets shown a lot of pages to page through, even though there are 0 results.

Nice catch. I just checked and this is also on _main_ - so not new - but still would be great to fix here!

corneliusroemer commented:
Indeed both are on prod, will make 2 issues and hide it here. Thanks @theosanderson

corneliusroemer commented on backend/docs/db/schema.sql:
```diff
@@ -468,12 +468,17 @@ CREATE VIEW public.sequence_entries_view AS
+            WHEN (sepd.processing_status = 'IN_PROCESSING'::text) THEN NULL::text
+            WHEN ((sepd.errors IS NOT NULL) AND (jsonb_array_length(sepd.errors) > 0)) THEN 'HAS_ERRORS'::text
+            WHEN ((sepd.warnings IS NOT NULL) AND (jsonb_array_length(sepd.warnings) > 0)) THEN 'HAS_WARNINGS'::text
+            ELSE 'NO_ISSUES'::text
```
Not blocking, should probably create issue but noticed here: it's rather strange we use a text column to store what is in reality an enum. We should use enums here: https://www.postgresql.org/docs/current/datatype-enum.html

corneliusroemer replied:
I've made a separate issue regarding possible schema improvements: https://github.com/loculus-project/loculus/issues/3201

fhennig replied:
thanks for picking this up and creating the ticket! I'm marking this as resolved here, as we won't do anything in this PR.

corneliusroemer commented on backend/docs/plantuml/sequenceInitialSubmission.puml:
```diff
@@ -21,7 +21,7 @@
         processing -> backend: Query initial/edited sequence data
         backend -> DB: Query initial/edited sequence data
-        backend -> DB: Set status from 'RECEIVED'/'HAS_ERRORS' to 'IN_PROCESSING'
+        backend -> DB: Set status from 'RECEIVED' to 'IN_PROCESSING'
```
I wonder whether the docs here were wrong in that one could also set status from "WAITING_FOR_APPROVAL" to "IN_PROCESSING", because that's what happened when one edited a sequence that passed warnings.

corneliusroemer replied:
Are we sure that status cannot go from 'PROCESSED' to 'IN_PROCESSING' here as well? Since we now call "HAS_ERRORS" "PROCESSED" this would be the logical change? So either this is a mistake or a bug fix of the original

corneliusroemer replied:
Never mind, it's just the old docs were wrong I think, this is easier to read and correct I think https://github.com/loculus-project/loculus/blob/5725b0b356529c788c6b3202cd9265ce8e9e79e8/backend/docs/plantuml/statusChange.svg

fhennig replied:
Yeah I think if an edit is done, the PreprocessedData is deleted and then it is 'RECEIVED', and then the pipeline picks it up again and eventually moves it to IN_PROCESSING again.

resolving this then!

corneliusroemer commented on backend/docs/plantuml/statusChange.puml:
```diff
@@ -18,17 +18,15 @@
-    state HAS_ERRORS
+    state PROCESSED
     state RECEIVED
     state APPROVED_FOR_RELEASE
```
While we're at it we could also rename "APPROVED_FOR_RELEASE" to "RELEASED", the former is unnecessarily complicated and wordy.

theosanderson replied:
(I support this change at some point, but agnostic as to whether in this PR)

corneliusroemer commented on backend/src/main/kotlin/org/loculus/backend/api/SubmissionTypes.kt:
```diff
@@ -291,10 +283,42 @@ enum class Status {
 }
 
+enum class ProcessingResult {
+    /**
+     * The sequence has no warnings or errors
+     */
```
```suggestion
    // The sequence has no warnings or errors
```

Are these docstrings auto-generated or have special meaning or why do we use 2 extra lines here 😄

corneliusroemer replied:
I'm not at all against comments (we currently have very few because of Fabian's preference), but adding extra white space for no reason is maybe overdoing it 🙃

corneliusroemer replied:
Ok, so ChatGPT told me these are KDoc comments, which are used by automatic tools, so that's a reason not to use `//` - however it seems possible to use `/** The sequence has no warnings or errors. */` instead here

fhennig replied:
[781edae](https://github.com/loculus-project/loculus/pull/3125/commits/781edae3a64c90385a3ae01ed85471bae68a4e30)

corneliusroemer commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -198,37 +203,24 @@ open class SubmissionDatabaseService(
-            val newStatus = insertProcessedDataWithStatus(submittedProcessedData, organism, pipelineVersion)
-
-            statusByAccessionVersion[submittedProcessedData.displayAccessionVersion()] = newStatus
+            insertProcessedDataWithStatus(submittedProcessedData, organism, pipelineVersion)
```
```suggestion
            insertProcessedDataWithStatus(submittedProcessedData, organism, pipelineVersion)
```

Does the "WithStatus" still make sense? we no longer do anything with the return value

corneliusroemer commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -198,37 +203,24 @@ open class SubmissionDatabaseService(
-
-            "Updated $finished sequences to $FINISHED, $hasErrors sequences to $HAS_ERRORS."
-        }
+        log.info("Updated ${processedAccessionVersions.size} sequences to $PROCESSED")
```
Slight loss of information here in the logs: we no longer know how many passed with and without errors, but probably not worth complexity to add this back

fhennig replied:
Added it back: [781edae](https://github.com/loculus-project/loculus/pull/3125/commits/781edae3a64c90385a3ae01ed85471bae68a4e30)

corneliusroemer commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -345,11 +332,9 @@ open class SubmissionDatabaseService(
     }
 
-    private fun postprocessAndValidateProcessedData(
+    private fun postProcessAndValidateProcessedData(
```
I'm not sure whether this change is for better or worse. I think a docstring here would really help as it's not immediately obvious from the name and arguments what this function does. It seems to do a status change, i.e. not be purse.

There might also be a misunderstanding. I _think_ this function name means postprocessing as a verb, not posting something, or doing something post=after "ProcessAndValidate"

corneliusroemer replied:
```suggestion
    private fun postprocessAndValidateProcessedData(
```

I think it's pretty clearly postprocessing that's meant, as the processed sequences are turned upper case among other things

corneliusroemer commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -313,17 +305,12 @@ open class SubmissionDatabaseService(
         organism: Organism,
         pipelineVersion: Long,
-    ): PreprocessingStatus {
+    ) {
```
Function should be renamed to remove the "WithStatus" as there's no longer any status emitted

fhennig replied:
https://github.com/loculus-project/loculus/pull/3125/commits/a72c1586cf148a28688cd4b3597a50377cf2a667

corneliusroemer commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -466,12 +451,13 @@ open class SubmissionDatabaseService(
                     .andThatOrganismIs(organism)
-                    .andThatSequenceEntriesAreInStates(listOf(Status.AWAITING_APPROVAL))
+                    .andThatSequenceEntriesAreInStates(listOf(Status.PROCESSED))
+                    .andThatSequenceEntriesHaveNoErrors()
```
This is a key change that we should test well - is this covered by a unit test? That if you try to approve a sequence with errors an error is thrown? It probably is covered, just double checking.

fhennig replied:
It actually wasn't covered. I agree that it's important to test. I added a test here:

[d5db67c](https://github.com/loculus-project/loculus/pull/3125/commits/d5db67c4b4cdcad892a6bbf716c5299e3074a746)

chaoran-chen commented on backend/src/main/resources/db/migration/V1.5__update_sequence_entries_view_status_field.sql:
For consistency, I think that we should write all SQL keywords in lower or upper case but not mix them (my preference is lower case).

fhennig replied:
agreed! I lower-cased it all.

[`f213a23` (#3125)](https://github.com/loculus-project/loculus/pull/3125/commits/f213a237c8b1f190741ff319ee17c4e4d047e2a5)

chaoran-chen commented on backend/src/test/kotlin/org/loculus/backend/controller/submission/SubmissionConvenienceClient.kt:
```diff
@@ -287,6 +287,24 @@ class SubmissionConvenienceClient(
+                ),
+        ).processingResultCounts[processingResult]!!.toInt()
+
     fun expectStatusCountsOfSequenceEntries(statusCounts: Map<Status, Int>, userName: String = DEFAULT_USER_NAME) {
```
IntelliJ tells me that this is not used anymore and can be deleted (this might not actually related to a change of this PR)

fhennig replied:
deleted https://github.com/loculus-project/loculus/pull/3125/commits/f213a237c8b1f190741ff319ee17c4e4d047e2a5

chaoran-chen commented on backend/src/main/resources/db/migration/V1.5__update_sequence_entries_view_status_field.sql:
```diff
@@ -0,0 +1,37 @@
+        se.accession = em.accession
+        and se.version = em.version;
+
+UPDATE sequence_entries_preprocessed_data
+SET processing_status = 'PROCESSED'
+WHERE processing_status IN ('HAS_ERRORS', 'FINISHED');
```
Indeed a great catch, @corneliusroemer, @anna-parker and @fhennig!

When we deploy a new version, Kubernetes will usually do a rolling update where it first starts the new backend and then shuts down the old one, right? (cc @theosanderson) If yes, this means that the old backend could still write "HAS_ERRORS" and "AWAITING_APPROVAL" into the database after the new backend executed the UPDATE query and thereby put the database into an invalid state.

chaoran-chen reviewed (commented):
Thanks for the great (and big) PR! I only managed to review the backend code again, I'll try to review the website tomorrow. I've a question regarding how we can safely deploy this, everything else is very minor.

fhennig commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -675,39 +670,77 @@ open class SubmissionDatabaseService(
+        authenticatedUser: AuthenticatedUser,
+        organism: Organism?,
+    ): Map<ProcessingResult, Int> {
+        val processingResultColum = SequenceEntriesView.processingResultColum
```
https://github.com/loculus-project/loculus/pull/3125/commits/f213a237c8b1f190741ff319ee17c4e4d047e2a5

fhennig commented on backend/docs/runtime_view.md:
```diff
@@ -41,12 +41,7 @@ When submitting a sequence entry, it goes through a series of statuses:
-The sequence entry was successfully processed by the preprocessing pipeline and can be released
-or a revocation was submitted.
-The submitter has to approve the release.
+**Processed**: The sequence entry has been processed by the preprocessing pipeline. It might now contain errors and needs to be edited and resubmitted, or it can be approved for release.
```
Oh yes 😅

fhennig commented on backend/docs/plantuml/statusChange.puml:
```diff
@@ -38,15 +36,12 @@
+    IN_PROCESSING --> PROCESSED : preprocessing done, might contain errors that need editing.
+    PROCESSED --> RECEIVED : user edits data
+    PROCESSED --> is_approved
     is_approved --> APPROVED_FOR_RELEASE : user approves \nor automatic approval\n via 'release_directly'
```
[feb8f09](https://github.com/loculus-project/loculus/pull/3125/commits/feb8f09b413f0a53c7a292817688e94cbceb9b43)

chaoran-chen commented on website/src/components/ReviewPage/ReviewPage.tsx:
```diff
@@ -149,50 +166,60 @@ const InnerReviewPage: FC<ReviewPageProps> = ({ clientConfig, organism, group, a
 
     const pagination = (
         <div className='flex justify-end align-center gap-3 py-3'>
             <Pagination
```
Something very small: if you first go to the last page and then filter out sequences, you end up on an empty page. Instead, I think that it would be better to go e.g. to the last of the existing pages:

https://github.com/user-attachments/assets/5d0d1f29-5de8-4860-aaf0-133dbfd4e110

But definitely feel free to move this to a new issue!

fhennig replied:
Should be fixed here: [979b7c0](https://github.com/loculus-project/loculus/pull/3125/commits/979b7c0d772a02ccd3d846c42dac819da0356ddd)

fhennig replied:
I didn't think about testing this, very good!!

chaoran-chen reviewed (approved):
Thanks, Felix! I think this is good to go!

corneliusroemer commented:
I wonder if this change should be reverted: https://github.com/loculus-project/loculus/pull/3125#discussion_r1846874164 until we have made it in code
